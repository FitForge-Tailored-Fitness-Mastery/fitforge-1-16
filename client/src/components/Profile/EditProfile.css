.edit-profile-container {
    max-width: 400px;
    min-width: 40vh;
    min-height: 100vh;
    margin: auto;
    padding: 20px;
    background-color: #333;
    color: white;
    font-family: Arial, sans-serif;
    border-radius: 10px;
  }
  
  .edit-profile-container h2 {
    text-align: center;
    margin-bottom: 20px;
  }
  
  .form-group {
    margin-bottom: 20px;
  }
  
  .form-group label {
    display: block;
    margin-bottom: 5px;
  }
  
  .form-group input {
    width: 70%;
    padding: 10px;
    border-radius: 5px;
    border: none;
  }
  
  button[type='submit'] {
    display: block;
    width: 100%;
    padding: 10px;
    background-color: gold;
    color: black;
    font-weight: bold;
    border: none;
    border-radius: 5px;
    cursor: pointer;
  }

  .edit-profile-container {
    max-width: 400px;
    margin: 0 auto;
    padding: 20px;
    background-color: #333;
    color: white;
    font-family: Arial, sans-serif;
    border-radius: 10px;
  }
  
  @media (max-width: 896px) {
    .profile-container {
      max-width: 90%;
      padding: 10px;
    }
  }
  
  @media (max-width: 480px) {
    .profile-container {
      margin: 20px 10px;
      padding: 10px;
    }
  }