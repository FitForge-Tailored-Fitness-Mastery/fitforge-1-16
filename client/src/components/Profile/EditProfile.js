import React, { useState, useEffect } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import NavigationBar from '../NavigationBar/NavigationBar';
import './EditProfile.css';

const EditProfile = () => {
  const { clientId } = useParams();
  const navigate = useNavigate();
  const [client, setClient] = useState(null);
  const [formData, setFormData] = useState({
    fname: '',
    lname: '',
    height: '',
    dob: '',
    weight: '',
  });
  const [error, setError] = useState(null); // Adding state to handle errors

  useEffect(() => {
    const fetchClientData = async () => {
      try {
        const response = await fetch(`http://localhost:5000/client/${clientId}`);
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`);
        }
        const data = await response.json();
        setClient(data);
        setFormData({
          fname: data.fname,
          lname: data.lname,
          height: data.height,
          dob: data.dob,
          weight: data.weight,
        });
      } catch (error) {
        console.error('Fetch error:', error);
        setError('Failed to load data');
      }
    };
    fetchClientData();
  }, [clientId]);

  const handleInputChange = (e) => {
    let { name, value } = e.target;
    console.log(e, name, value)
    if (name === 'dob') {
      const date = new Date(value);
      value = date.toISOString().slice(0, 10); // Converts date to YYYY-MM-DD format
      console.log('Formatted Date:', value); // Check the formatted date
    }
    setFormData({ ...formData, [name]: value });
  };
    

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const response = await fetch(`http://localhost:5000/client/${clientId}`, {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(formData),
      });
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      const data = await response.json();
      console.log('Client updated:', data);
      navigate('/profile'); // Navigate back to the profile page after successful update
    } catch (error) {
      console.error('Update error:', error);
      setError(`Failed to update data: ${error.message}`); // Include the error message in the state
    }
  };  

  if (!client) {
    return <div>Loading...</div>;
  }

  return (
    <div className="edit-profile-container">
      <div className='edit-profile-box'>
          <h2>Edit Profile</h2>
          {/* Displaying error message */}
          {error && <p className="error">{error}</p>}
          <form id='editprofileform' onSubmit={handleSubmit}>
            <div className="form-group">
              <label htmlFor="fname">First Name:</label>
              <input
                type="text"
                id="fname"
                name="fname"
                value={formData.fname}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="lname">Last Name:</label>
              <input
                type="text"
                id="lname"
                name="lname"
                value={formData.lname}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="height">Height (cm):</label>
              <input
                type="number"
                id="height"
                name="height"
                value={formData.height}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="dob">Date of Birth:</label>
              <input
                type="date"
                id="dob"
                name="dob"
                value={formData.dob}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="weight">Weight (lb):</label>
              <input
                type="number"
                id="weight"
                name="weight"
                value={formData.weight}
                onChange={handleInputChange}
              />
            </div>
            <button type="submit" className='editprofilebutton'>Save Changes</button>
          </form>
      </div>
      {/* Add Navigation Bar */}
      <NavigationBar />
    </div>
  );
};

export default EditProfile;
