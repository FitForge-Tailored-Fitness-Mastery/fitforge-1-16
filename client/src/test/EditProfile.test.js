import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import EditProfile from '../components/Profile/EditProfile';
import { BrowserRouter } from 'react-router-dom';
import { useParams } from 'react-router-dom';

const mockedNavigate = jest.fn();
// Mock useParams
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
  useNavigate: () => mockedNavigate,
}));

describe('EditProfile', () => {
  beforeEach(() => {
    global.fetch = jest.fn();
    useParams.mockReturnValue({ clientId: '123' });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('displays loading state initially', async () => {
    global.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        imageUrl: 'http://example.com/image.jpg',
        fname: 'John',
        lname: 'Doe',
        email: 'john.doe@example.com',
        phone: '123-456-7890',
      }),
    });
    const { asFragment } = render(<BrowserRouter><EditProfile /></BrowserRouter>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders all form fields and a submit button', async () => {
    global.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        imageUrl: 'http://example.com/image.jpg',
        fname: 'John',
        lname: 'Doe',
        email: 'john.doe@example.com',
        phone: '123-456-7890',
      }),
    });
    render(<BrowserRouter><EditProfile /></BrowserRouter>);

    // Simulate loading data
    await waitFor(() => {
      expect(screen.getByLabelText('First Name:')).toBeInTheDocument();
      expect(screen.getByLabelText('Last Name:')).toBeInTheDocument();
      expect(screen.getByLabelText('Height (cm):')).toBeInTheDocument();
      expect(screen.getByLabelText('Date of Birth:')).toBeInTheDocument();
      expect(screen.getByLabelText('Weight (lb):')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Save Changes' })).toBeInTheDocument();
    });
  });

  test('fills out and submits the form', async () => {
    global.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        imageUrl: 'http://example.com/image.jpg',
        fname: 'John',
        lname: 'Doe',
        email: 'john.doe@example.com',
        phone: '123-456-7890',
      }),
    });
    render(<BrowserRouter><EditProfile /></BrowserRouter>);

    await waitFor(() => {
      const firstNameInput = screen.getByLabelText('First Name:');
      fireEvent.change(firstNameInput, { target: { value: 'Jane' } });

      const submitButton = screen.getByRole('button', { name: 'Save Changes' });
      fireEvent.click(submitButton);
    });

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith(expect.any(String), expect.objectContaining({
        method: 'PATCH',
        body: "{\"fname\":\"Jane\",\"lname\":\"Doe\"}",
        headers: {'Content-Type': 'application/json'},
      }));
    })
  });
});
