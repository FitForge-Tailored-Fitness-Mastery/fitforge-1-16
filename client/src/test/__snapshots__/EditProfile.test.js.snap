// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfile displays loading state initially 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
